sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -3198,9 +3199,66 @@ class Dataset[T] private[sql](
   }
 
   /**
-   * Collect a Dataset as Arrow batches and serve stream to PySpark.
+   * Collect a Dataset as Arrow batches and serve stream to SparkR. It sends
+   * arrow batches in an ordered manner with buffering. This is inevitable
+   * due to missing R API that reads batches from socket directly. See ARROW-4512.
+   * Eventually, this code should be deduplicated by `collectAsArrowToPython`.
    */
-  private[sql] def collectAsArrowToPython(): Array[Any] = {
+  private[sql] def collectAsArrowToR(): Array[Any] = {
```

Comment:
Can we unit test this in Scala tests?

Reply:
Yes, this was pointed out by @BryanCutler at https://github.com/apache/spark/pull/23760#issuecomment-463012213.

This `collectAsArrowToR` was restored from the previous code (see bf2feec). The newer code (for `collectAsArrowToPython`) has a test by using Python's RDD APIs.

If I am not mistaken, it's difficult to write a test with R RDD API, and we're in transition to get rid of RDD API in SparkR completely (https://github.com/apache/spark/pull/22866).

Since this was restored almost as is from the previous code, and eventually, this code will be deduplicated by `collectAsArrowToPython` (after Arrow 0.14.0 release, this logic is restored mainly due to ARROW-4512), I hope end-to-end tests only are okay enough.

Since it's being tested via Python RDD API, we could deduplicate the tests too after Arrow 0.14.0 release. Does that make sense to you?

Reply:
> Can we unit test this in Scala tests?

It would be difficult to test in Scala because this method sets up a server in the JVM for the R process to read data from.

@HyukjinKwon maybe you can skip the part of the test in https://github.com/apache/spark/commit/bf2feec2ef023177d72ac1137dbd1b3a02eb9a89 that uses the delay with the RDD and just add the case that uses a large number of partitions. That will make a pretty good chance to test partitions coming in out of order.

Reply:
Yup, let me add a test. Thanks, @BryanCutler